src/packet_analysis/protocol/ip/IPBasedAnalyzer.h

```diff
@@ -0,0 +1,182 @@
+	 * responder based on known ports and such.
+	 *
+	 * @param src_port The source port of the connection.
+	 * @param dst_port The destination port of the connection.
+	 * @param data The payload data for the packet being processed.
+	 * @param flip_roles Return value if the roles should be flipped.
+	 * @return True if the connection is wanted. False otherwise.
+	 */
+	virtual bool WantConnection(uint16_t src_port, uint16_t dst_port,
+	                            const u_char* data, bool& flip_roles) const
+		{
+		flip_roles = false;
+		return true;
+		}
+
+	/**
+	 * Returns the port mask for an analyzer. Used by IsLikelyServerPort().
+	 */
+	virtual uint32_t GetServerPortMask() const = 0;
+
+	/**
+	 * Returns the transport protocol. Used by NewConn().
+	 */
+	virtual TransportProto GetTransportProto() const = 0;
+
+	/**
+	 * Returns true if the port corresponds to an application for which there
+	 * is a Zeek analyzer (even if it might not be used by the present policy
+	 * script) or if it's generally a likely server port.
+	 *
+	 * @param port The port number to check, in host order.
+	 */
+	bool IsLikelyServerPort(uint32_t port) const;
+
+	/**
+	 * Continues process of packet after the connection has been inserted into the
+	 * session manager. This should be implemented by all child classes.
+	 *
+	 * @param conn The connection currently being processed.
+	 * @param t The timestamp for the current packet.
+	 * @param is_orig Flag denoting whether this packet is from the originator of
+	 * the connection.
+	 * @param remaining The remaining about of data in the packet.
+	 * @param pkt The packet being processed.
+	 */
+	virtual void ContinueProcessing(Connection* conn, double t, bool is_orig, int remaining,
+	                                Packet* pkt) {}
+
+	/**
+	 * The entry point to the session analysis framework. This should be reset between calls
+	 * to ContinueProcessing(), as a packet analyzer should not maintain any state between
+	 * packets.
+	 */
+	IPBasedTransportAnalyzer* session_analyzer;
+
+	// TODO: temporary, until all of the plugins are implemented
+	bool new_plugin = false;
+
+private:
+
+	/**
+	 * Creates a new Connection object from data gleaned from the current packet.
+	 *
+	 * @param id A connection ID generated from the packet data. This should have been
+	 * passed in from a child analyzer.
+	 * @param key A connection ID key generated from the ID.
+	 * @param pkt The packet associated with the new connection.
+	 */
+	zeek::Connection* NewConn(const ConnTuple* id, const detail::ConnKey& key,
+	                          const Packet* pkt);
+};
+
+/**
+ * This class represents the interface between the packet analysis framework and
+ * the session analysis framework. One of these should be implemented for each
+ * packet analyzer that intends to forward into the session analysis.
+ */
+class IPBasedTransportAnalyzer : public zeek::analyzer::TransportLayerAnalyzer {
```

Comment:
I'm wondering if this should move into `analyzer/`. That would fit semantically, and it would put into the analyzer namespace as well, which would make it quite a bit more readable I think. Right now, we have `packet_analysis::IP::IPBasedAnalyzer`, which is a packet analyzer; and then `packet_analysis::IP::IPBasedTransportAnalyzer`, which is a connection analyzer -- that's feels to track.

The next question then would be if should split up UDP/ICMP/TCP similarly. I'm inclined to say yes, but not fully sure.

Reply:
I think the problem here is really naming and not location. The "transport analyzers" here aren't really analyzers in the pure sense. They're really just a glue layer between packet analysis and the child analyzers that used to be in the old analyzers. They're not intended to be called by anyone other than the packet analyzers they're associated with, so I'm sort of leery to move them.

Reply:
Ok, I can see this being primarily an issue of naming.